### ISSUE TITLE
arm-unknown-linux-gnueabihf, gets Illegal Instruction when run on RPI Zero W

### ISSUE THREAD

**Issue by tmayoff** [cross-rs/cross#1017]
### Checklist

- [X] I've looked through the [issues and pull requests](https://github.com/cross-rs/cross/issues?q=) for similar reports

### Describe your issue

I'm trying to cross compile for my RPI Zero W, and just get Illegal Instruction when I try to run it. I cut the code down to a simple hello world, and still get Illegal Instruction

This is the backtrace produced under gdb.
```
#0  0x0054640c in __libc_csu_init ()
#1  0xb6da06cc in __libc_start_main (main=0xbefff654, argc=-1225945088, argv=0xb6da0690 <__libc_start_main+100>, init=0x54640d <__libc_csu_init>, fini=0x54644d <__libc_csu_fini>, rtld_fini=0xb6fdd510 <_dl_fini>, stack_end=0xbefff654)
    at libc-start.c:264
#2  0x00406444 in _start () at ../ports/sysdeps/arm/start.S:110
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```
Any help would be appreciated

### What target(s) are you cross-compiling for?

arm-unknown-linux-gnueabihf

### Which operating system is the host (e.g computer cross is on) running?

- [ ] macOS
- [ ] Windows
- [X] Linux / BSD
- [ ] other OS (specify in description)

### What architecture is the host?

- [X] x86_64 / AMD64
- [ ] arm32
- [ ] arm64 (including Mac M1)

### What container engine is cross using?

- [ ] docker
- [X] podman
- [ ] other container engine (specify in description)

### cross version

cross 0.2.4 (47df5c7 2022-08-30)

**Comment by Alexhuszagh** [comment]
This should be linked to #426. I wonder if our update to fix targeting ARMv6 HF platforms was insufficient. Previously, we erroneously compiled to an ARMv7 target, which would have caused the bug.

This should have been changed in the v0.2.4 branch, but just to check: can you run `readelf -A $binary`, where `$binary` is the path to the cross-compiled hello world binary? It should output ARMv6 if it is, in fact, using the correct toolchain. If not, try updating the toolchain via `docker pull`. Let me know if you have any questions.

**Comment by tmayoff** [comment]
```
Attribute Section: aeabi
File Attributes
  Tag_CPU_name: "7-A"
  Tag_CPU_arch: v7
  Tag_CPU_arch_profile: Application
  Tag_ARM_ISA_use: Yes
  Tag_THUMB_ISA_use: Thumb-2
  Tag_FP_arch: VFPv3-D16
  Tag_ABI_PCS_GOT_use: GOT-indirect
  Tag_ABI_PCS_wchar_t: 4
  Tag_ABI_FP_rounding: Needed
  Tag_ABI_FP_denormal: Needed
  Tag_ABI_FP_exceptions: Needed
  Tag_ABI_FP_number_model: IEEE 754
  Tag_ABI_align_needed: 8-byte
  Tag_ABI_align_preserved: 8-byte, except leaf SP
  Tag_ABI_enum_size: int
  Tag_ABI_VFP_args: VFP registers
  Tag_CPU_unaligned_access: v6
  Tag_ABI_FP_16bit_format: IEEE 754
```

**Comment by Alexhuszagh** [comment]
> ```
> Attribute Section: aeabi
> File Attributes
>   Tag_CPU_name: "7-A"
>   Tag_CPU_arch: v7
>   Tag_CPU_arch_profile: Application
>   Tag_ARM_ISA_use: Yes
>   Tag_THUMB_ISA_use: Thumb-2
>   Tag_FP_arch: VFPv3-D16
>   Tag_ABI_PCS_GOT_use: GOT-indirect
>   Tag_ABI_PCS_wchar_t: 4
>   Tag_ABI_FP_rounding: Needed
>   Tag_ABI_FP_denormal: Needed
>   Tag_ABI_FP_exceptions: Needed
>   Tag_ABI_FP_number_model: IEEE 754
>   Tag_ABI_align_needed: 8-byte
>   Tag_ABI_align_preserved: 8-byte, except leaf SP
>   Tag_ABI_enum_size: int
>   Tag_ABI_VFP_args: VFP registers
>   Tag_CPU_unaligned_access: v6
>   Tag_ABI_FP_16bit_format: IEEE 754
> ```
> 
> Also I should note I'm using a custom Docker image based on this `FROM ghcr.io/cross-rs/arm-unknown-linux-gnueabihf:edge`

Can you try updating the base image or rebuilding the custom image and trying again? Something like:

```bash
docker pull ghcr.io/cross-rs/arm-unknown-linux-gnueabihf:edge
docker build --no-cache ...
```

And then retrying a clean build? I'm assuming you're using an older image, since this should be fixed in a newer image. If you're linking any external libraries, however, notice you can't use system packages since the `armel` Debian architecture is ARMv7. For an example of a working hello world example, we have:

```bash
$ cargo init --bin hello
     Created binary (application) package
$ cd hello/
$ docker pull ghcr.io/cross-rs/arm-unknown-linux-gnueabihf:edge
$ CROSS_TARGET_ARM_UNKNOWN_LINUX_GNUEABIHF_IMAGE="ghcr.io/cross-rs/arm-unknown-linux-gnueabihf:edge" cross build --target arm-unknown-linux-gnueabihf
$ readelf -A target/arm-unknown-linux-gnueabihf/debug/hello
Attribute Section: aeabi
File Attributes
  Tag_CPU_name: "6"
  Tag_CPU_arch: v6
  Tag_ARM_ISA_use: Yes
  Tag_THUMB_ISA_use: Thumb-1
  Tag_FP_arch: VFPv2
  Tag_ABI_PCS_GOT_use: GOT-indirect
  Tag_ABI_PCS_wchar_t: 4
  Tag_ABI_FP_rounding: Needed
  Tag_ABI_FP_denormal: Needed
  Tag_ABI_FP_exceptions: Needed
  Tag_ABI_FP_number_model: IEEE 754
  Tag_ABI_align_needed: 8-byte
  Tag_ABI_enum_size: int
  Tag_ABI_VFP_args: VFP registers
  Tag_CPU_unaligned_access: v6
  Tag_ABI_FP_16bit_format: IEEE 754
```

If you notice, your example compiles to ARMv7, which was a bug prior to the fix in #734, but the latest image compiles to ARMv6. If you have no external dependencies, updating your cross image and custom image **should** fix this.

**Comment by Alexhuszagh** [comment]
Oh one quick way you can see how old an image is is by running the following:

```bash
$ docker image inspect ghcr.io/cross-rs/arm-unknown-linux-gnueabihf:edge | grep "org.opencontainers.image.created"
                "org.opencontainers.image.created": "2022-09-08T13:49:50.077Z",
```

So, my local image is from Sept 8th, but (I am guessing) your image is a bit older.

**Comment by tmayoff** [comment]
Okay so the image is up to date, I just setup this up today output is ` "org.opencontainers.image.created": "2022-09-08T13:49:50.077Z"`. 

I do depend on some c libraries, lirc-client and sqlite3

```
If you're linking any external libraries, however, notice you can't use system packages since the armel Debian architecture is ARMv7
```
Not sure I understand what you mean here?

**Comment by tmayoff** [comment]
So with an actual hello world test it works

**Comment by Alexhuszagh** [comment]
> do depend on some c libraries, lirc-client and sqlite3

So this is the issue: how did you install these libraries to link them? Did you compile them from source? Or, as I suspect, did you do something like:

```bash
$ sudo dpkg --add-architecture armhf
$ sudo apt-get update 
$ sudo apt-get -y install libsqlite3-dev:armhf
```

And then use that to build your package? Because this will fail, since the `armhf` libraries are built for an ARMv7 architecture, not ARMv6. So you cannot use `apt` to install your packages: you must use something like `conan`, `vcpkg`, or build them from source.

**Comment by tmayoff** [comment]
I'll try compiling the dependencies from source, is it enough to just compile them from within the docker image, or do I need get the proper toolchain for compiling it as well

**Comment by Alexhuszagh** [comment]
> I'll try compiling the dependencies from source, is it enough to just compile them from within the docker image, or do I need get the proper toolchain for compiling for armv6 as well

You should build them as part of building your Docker image. You can use the toolchain provided in the base cross image to do so. So, for example, the C compiler is `arm-unknown-linux-gnueabihf-gcc`, the C++ compiler is `arm-unknown-linux-gnueabihf-g++`, the AR is `arm-unknown-linux-gnueabihf-ar`, etc. Just use those values for your build configuration (they will likely need to be provided to make/cmake/etc.), which may require a toolchain file or some external defines.

Are you at all familiar with C and C++ build systems? If not, I highly recommend opening a discussion or joining our Matrix room: it's late here tonight but I'd be more than happy to walk you through the process.

**Comment by tmayoff** [comment]
I am much more familiar with C/C++ and their build systems than with rust, I think I'll be fine for now thank you very much for the help

**Comment by Alexhuszagh** [comment]
Wonderful, good to hear. Let me know if anything seems unusual or I can help in any way.

**Comment by Alexhuszagh** [comment]
@tmayoff FYI: this isn't the first time this issue has occurred, so I'm thinking about a more permanent solution. We can't really detect if the custom image or build script logic you use does something you shouldn't (install an `armhf` architecture package), but we can denylist every single `armhf` package and so any derivative images or build scripts with faulty logic will fail to find an installation candidate for those packages.

In short, we could make it so if you try to install `sqlite3:armhf`, it will say no package was found (unless you really know what you're undoing and undo our package pins). Does that sound like a good solution? Idea is in #1018.
